lib/pgp.ml

```diff
@@ -537,24 +540,216 @@ module Packet = struct
+
+      type t =
+        | Key_expiration_time of int32
+        | Revocation_reason of string
+        | Issuer_id of string
+        | Uses of key_flag list
+        | Unknown
+      [@@deriving ord, eq, show]
+
+      let get_length cs =
+        let first_octet = Cstruct.get_uint8 cs 0 in
+        if first_octet < 192 then
+          (1, first_octet)
+        else if first_octet < 255 then
+          let second_octet = Cstruct.get_uint8 cs 1 in
+          let length = 192 + second_octet + (256 * (first_octet - 192)) in
+          (2, length)
+        else
+          let length = Cstruct.BE.get_uint32 cs 1 in
+          (5, Int32.to_int length)
+
+      let rec keep_some opt_list =
+        match opt_list with
+        | Some x :: _list -> x :: keep_some _list
+        | None :: _list -> keep_some _list
+        | [] -> []
+
+      let decode_uses cs =
+        let nth_bit x n = x land (1 lsl n) <> 0 in
+        let flag_int = Cstruct.get_uint8 cs 1 in
+        let map_tag tag flag =
+          if flag then
+            Some (tag_to_keyflag tag)
+          else
+            None
+        in
+        let flags = List.init 8 (nth_bit flag_int) in
+        if List.nth flags 6 then
+          Error (Printf.sprintf "Bad tag for a key flag: 0x40")
+        else
+          Ok (Uses (keep_some (List.mapi map_tag flags)))
+
+      let decode_code_reason code =
+        match code with
+        | 0 -> Ok ""
+        | 1 -> Ok "Key is superseded: "
+        | 2 -> Ok "Key material has been compromised: "
+        | 3 -> Ok "Key is retired and no longer used: "
+        | i when 100 <= i && i <= 110 -> Ok ""
+        | 32 -> Ok "User ID information is no longer valid: "
+        | code -> Error (Printf.sprintf "Bad revocation reason code: %i" code)
+
+      let decode_revocation_reason cs =
+        let code = Cstruct.get_uint8 cs 1 in
+        decode_code_reason code >|= fun reason_code_string ->
+        let reason_string = Cstruct.to_string ~off:2 cs in
+        Revocation_reason (reason_code_string ^ reason_string)
+
+      let decode_subpacket cs =
+        let type_code = Cstruct.get_uint8 cs 0 in
+        match type_code with
+        | 9 -> Ok (Key_expiration_time (Cstruct.BE.get_uint32 cs 1))
+        | 16 ->
+          let id = Cstruct.BE.get_uint64 cs 1 in
+          Ok (Issuer_id (Int64.format "%x" id))
+        | 27 -> decode_uses cs
+        | 29 -> decode_revocation_reason cs
+        | _ -> Ok Unknown
+
+      let rec decode_rec cs subpacket_list =
+        if Cstruct.length cs != 0 then
+          let (offset, length) = get_length cs in
+          let next_cs = Cstruct.shift cs (length + offset) in
+          let subpacket_cs = Cstruct.sub cs offset length in
+          decode_subpacket subpacket_cs >>= fun subpacket ->
+          decode_rec next_cs (subpacket :: subpacket_list)
+        else
+          Ok (List.rev subpacket_list)
+
+      let decode cs =
+        let filter_unknown_subpacket = function
```

Comment:
I would rename this into `is_filter_unknown_subpacket`. More generally, functions that are tests should always be called `is_...`, this way it is easy to identify them.